### ISSUE TITLE
[prefer_collection_literals] false positive.

### ISSUE THREAD

**Issue by modulovalue** [dart-lang/linter#4821]
Consider:

```dart
import 'dart:collection';

class Foo {
  LinkedHashMap<int, double> a;
  
  // Unnecessary constructor invocation. [prefer_collection_literals]
  //          vvvvvvvvvvvvvvv
  Foo() : a = LinkedHashMap();
}
```

If we prefer a collection literal, the code will break. Therefore, this is not an unnecessary constructor invocation. We can't prefer a collection literal here.

**Comment by lrhn** [comment]
While it's correct that the code won't work, you *really* shouldn't be using `LinkedHashMap` as a type. It's an implementation of `Map` and has exactly the same API, so just use `Map` as the type, and only use `LinkedHashMap` when calling the constructor. (And then it's correct that you should use the literal instead.)

**Comment by modulovalue** [comment]
> you really shouldn't be using LinkedHashMap as a type. It's an implementation of Map and has exactly the same API, so just use Map as the type

By using `LinkedHashMap` as a type we can be explicit about the fact that we need it to be a **linked** hash map and not just any potentially unlinked hash map. This is an implementation detail that can matter, and by using LinkedHashMap (or LinkedHashSet) as a type we can be explicit about the fact that it does matter.

**Comment by srawlins** [comment]
Fixed in https://github.com/dart-lang/sdk/commit/e1bec8d7d6802dbb61b6068b09f3c54b87770e9c
